File: url-shortener/backend/internal/domain/analytics.go
package domain

type AnalyticsOverview struct {
	TotalClicks   int64   `json:"total_clicks"`
	UniqueClicks  int64   `json:"unique_clicks"`
	BounceRate    float64 `json:"bounce_rate"`
	AvgTimeOnSite float64 `json:"avg_time_on_site"`
}

type ClicksOverTime struct {
	Labels []string `json:"labels"`
	Values []int    `json:"values"`
}

type LocationStat struct {
	Country     string  `json:"country"`
	CountryCode string  `json:"country_code"`
	Percent     float64 `json:"percent"`
}

type DeviceStat struct {
	Name    string  `json:"name"`
	Value   int     `json:"value"`
	Percent float64 `json:"percent"`
	Color   string  `json:"color"`
}

type BounceRateStat struct {
	LinkURL       string  `json:"link_url"`
	BounceRate    float64 `json:"bounce_rate"`
	TotalSessions int     `json:"total_sessions"`
}

type TrackSessionRequest struct {
	SessionID   string `json:"session_id"`
	LinkCode    string `json:"link_code"`
	UserIPHash  string `json:"user_ip_hash"`
	CountryCode string `json:"country_code"`
	DeviceType  string `json:"device_type"`
	BrowserName string `json:"browser_name"`
}

File: url-shortener/frontend/src/services/analytics.ts
import axios from 'axios'

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080/api'

// Типы для ответов бэкенда
export interface AnalyticsOverview {
  total_clicks: number
  unique_clicks: number
  bounce_rate: number
  avg_time_on_site: number
}

export interface ClicksOverTime {
  labels: string[]
  values: number[]
}

export interface LocationStat {
  country: string
  country_code: string
  percent: number
}

export interface DeviceStat {
  name: string
  value: number
  percent: number
  color: string
}

export interface BounceRateStat {
  link_url: string
  bounce_rate: number
  total_sessions: number
}

export const analyticsAPI = {
  async getOverview(days: number = 7): Promise<AnalyticsOverview> {
    const response = await axios.get(`${API_URL}/analytics/overview`, {
      params: { days },
      headers: { 'Authorization': `Bearer ${localStorage.getItem('access_token')}` }
    })
    // Защита от null
    return response.data || { total_clicks: 0, unique_clicks: 0, bounce_rate: 0, avg_time_on_site: 0 }
  },

  async getClicksOverTime(days: number = 7): Promise<ClicksOverTime> {
    const response = await axios.get(`${API_URL}/analytics/clicks-over-time`, {
      params: { days },
      headers: { 'Authorization': `Bearer ${localStorage.getItem('access_token')}` }
    })
    return response.data || { labels: [], values: [] }
  },

  async getTopLocations(days: number = 7, limit: number = 5): Promise<LocationStat[]> {
    const response = await axios.get(`${API_URL}/analytics/top-locations`, {
      params: { days, limit },
      headers: { 'Authorization': `Bearer ${localStorage.getItem('access_token')}` }
    })
    return response.data || []
  },

  async getDeviceStats(days: number = 7): Promise<DeviceStat[]> {
    const response = await axios.get(`${API_URL}/analytics/devices`, {
      params: { days },
      headers: { 'Authorization': `Bearer ${localStorage.getItem('access_token')}` }
    })
    return response.data || []
  },

  async getBounceRate(days: number = 7): Promise<BounceRateStat[]> {
    const response = await axios.get(`${API_URL}/analytics/bounce-rate`, {
      params: { days },
      headers: { 'Authorization': `Bearer ${localStorage.getItem('access_token')}` }
    })
    return response.data || []
  }
}